[CI] Rework github workflow processing

Keenuts opened the pull request:
Before this patch, the job/workflow name impacted the metric name, meaning a change in the workflow definition could break monitoring. This patch adds a map to get a stable name on metrics from a workflow name.

In addition, it reworks a bit how we track the last processed workflow: the github queries are broken if filtering is applied, meaning we have a list of workflow, ordered by 'created_at', which mixes completed & running workflows.
We have no guarantees over the order of completion, meaning we cannot stop at the first completed job we found (even per-workflow).

This PR processed the last 1000 workflows, but allows an early stop if the created_at time is older than 8 hours. This means we could miss long-running workflows (>8 hours), and if the number of workflows started before another one completes becomes high (>1000), we'll miss it.
To detect this kind of behavior, a new metric is added "oldest workflow processed", which should at least indicate if the depth is too small. 

An alternative without arbitrary cut would be to initially parse all workflows, and then record the last non-completed one we find and always start from the last (moving the lower bound as they complete).  But LLVM has forever-queued workflows runs (>1 years), hence this would cause us to iterate over a very large number of jobs.

Keenuts committed:
[CI] Rework github workflow processing

Before this patch, the job/workflow name impacted the metric name,
meaning a change in the workflow definition could break monitoring.
This patch adds a map to get a stable name on metrics from a workflow
name.

In addition, it reworks a bit how we track the last processed workflow
to simplify the behavior, and work around an API issue which returns
bogus results if a filter is used.

This PR is a first step to bring buildkite metrics monitoring.

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts committed:
format

boomanaiden154 commented on .ci/metrics/metrics.py:
```diff
@@ -12,9 +13,29 @@
-GITHUB_PROJECT = "llvm/llvm-project"
-WORKFLOWS_TO_TRACK = ["LLVM Premerge Checks"]
-SCRAPE_INTERVAL_SECONDS = 5 * 60
+SCRAPE_INTERVAL_SECONDS = 60
```
I'm assuming you've tested for this one hour plus to ensure we don't hit any API rate limits?

Keenuts replied:
Forgot to revert this, used 60s locally for testing, but 5mn is enough for what we need.

boomanaiden154 commented on .ci/metrics/metrics.py:
```diff
@@ -45,131 +65,116 @@ def get_sampled_workflow_metrics(github_repo: github.Repository):
     Returns:
-      Returns a list of JobMetrics objects, containing the relevant metrics about
-      the workflow.
+      Returns a tuple with 2 elements:
```
Can you add this as a type annotation to the function too?

boomanaiden154 commented on .ci/metrics/metrics.py:
```diff
@@ -191,25 +196,23 @@ def get_per_workflow_metrics(
             # The timestamp associated with the event is expected by Grafana to be
             # in nanoseconds.
-            created_at_ns = int(created_at.timestamp()) * 10**9
+            completed_at_ns = int(completed_at.timestamp()) * 10**9
```
This should be better than what we're doing now, but isn't guaranteed to not try and upload historical information. I think going by the workflow completion time should fix that, but I couldn't a good way to get that out of the workflow, other than just iterating through all the jobs.

Keenuts replied:
Yep, added a condition to check the metric age before adding it, thus dropping metrics older than 2 hours.

boomanaiden154 reviewed (commented):
Some comments, mostly minor. This looks pretty good.

Thanks for taking a stab at this!

Keenuts committed:
iterate over fixed depth

Keenuts committed:
pr-feedback

Keenuts commented:
Updated the description. realized during the weekend the method was flawed, and ended-up doing something even more basic: fetch the last 1000 workflows, and extract the metrics from those.
Added a cut-off for 8 hours old workflows, and also logging the max age oldest workflow we processed to monitor in case one day we have more then 1000 workflows in the last 2 hours.

boomanaiden154 commented on .ci/metrics/metrics.py:
```diff
@@ -178,38 +160,68 @@ def get_per_workflow_metrics(
+            metric_age_mn = (
+                datetime.datetime.now(datetime.timezone.utc) - completed_at
+            ).total_seconds() / 60
+            if metric_age_mn > GRAFANA_METRIC_MAX_AGE_MN:
```
Can you add some logging here so we can see better when this happens?

Keenuts replied:
added logging

boomanaiden154 reviewed (approved):
One nit, otherwise the new approach makes sense to me.

boomanaiden154 commented:
I didn't even consider the case where we hit a completed workflow that shadowed other completed workflows due to them being ordered by start time. That's a good catch.

I'm reasonably hopeful this will fix the weird issues we've been seeing. It looks like we may need to increase the max depth though based on what I saw on the dashboard today.

Keenuts committed:
add logging when dropping stale metrics

Keenuts commented:
> I didn't even consider the case where we hit a completed workflow that shadowed other completed workflows due to them being ordered by start time. That's a good catch.


Yes, I realized that very late.. So we probably had better metrics than reality because of this.

> It looks like we may need to increase the max depth though based on what I saw on the dashboard today.

Yes, seems like yesterday was a pretty busy day, but we might want to double it soon.
